extension/goplugin/util.go

```
@@ -136,34 +234,148 @@ func (util *Util) ResourceToMap(resource interface{}) map[string]interface{} {
+			fieldsMap[fieldName] = id
 			v.SetString(id)
-		} else if strings.Contains(v.Type().String(), "goext.Null") {
-			valid := v.FieldByName("Valid").Bool()
-			if valid {
-				fieldsMap[field] = v.FieldByName("Value").Interface()
-			} else {
-				fieldsMap[field] = nil
-			}
+		} else if isMaybeType(v.Kind(), v.Type().String()) {
+			maybeToMap(fieldsMap, fieldName, v)
 		} else if v.Kind() == reflect.Ptr {
 			if v.IsNil() {
-				fieldsMap[field] = nil
+				fieldsMap[fieldName] = nil
 			} else {
-				fieldsMap[field] = util.ResourceToMap(val)
+				rv := util.ResourceToMap(val)
+				fieldsMap[fieldName] = rv
 			}
+		} else if v.Kind() == reflect.Slice {
+			util.sliceToMap(fieldsMap, fieldName, v)
+		} else if v.Kind() == reflect.Struct {
+			fieldsMap[fieldName] = util.ResourceToMap(v.Addr().Interface())
 		} else {
-			fieldsMap[field] = val
+			fieldsMap[fieldName] = val
 		}
 	}
 
 	return fieldsMap
 }
+
+func (util *Util) sliceToMap(fieldsMap map[string]interface{}, fieldName string, v reflect.Value) {
+	if v.IsNil() {
+		// nothing
+	} else {
+		sliceElem := v.Type().Elem()
+		if isPrimitiveKind(sliceElem.Kind()) {
+			slice := make([]interface{}, v.Len())
+			for i := 0; i < v.Len(); i++ {
+				slice[i] = v.Index(i).Interface()
+			}
+			fieldsMap[fieldName] = slice
+		} else {
+			slice := make([]map[string]interface{}, v.Len())
+			for i := 0; i < v.Len(); i++ {
+				elem := v.Index(i)
+				if isNullableKind(elem.Kind()) {
+					if !elem.IsNil() {
+						slice[i] = util.ResourceToMap(elem.Interface())
+					} else {
+						slice[i] = nil
+					}
+				} else if elem.Kind() == reflect.Struct {
+					slice[i] = util.ResourceToMap(elem.Addr().Interface())
+				}
+			}
+			fieldsMap[fieldName] = slice
+		}
+	}
+}
+
+func maybeToMap(fieldsMap map[string]interface{}, fieldName string, value reflect.Value) {
+	if value.Kind() == reflect.Ptr {
+		if !value.IsNil() {
+			value = value.Elem()
+		} else {
+			// nothing, undefined value shoudn't appear in map
```

I'd prefer to add panic here instead silently ignoring the problem